File: src/api/daemon/encryption.ts
/**
 * Encryption API module — typed accessors for daemon encryption endpoints.
 *
 * 加密 API 模块 — daemon 加密端点的类型化访问器。
 *
 * # Endpoints / 端点
 * - `GET /encryption/state` → current encryption initialization & session state
 * - `POST /encryption/unlock` → auto-unlock encryption session via keyring (no passphrase)
 * - `POST /encryption/unlock-with-passphrase` → user-driven passphrase unlock (ADR-008 D15)
 * - `POST /encryption/lock` → lock encryption session (clear master key)
 * - `POST /encryption/factory-reset` → wipe key material + clear setup status
 * - `GET /encryption/keychain-access` → verify Keychain "Always Allow" permission
 */

import {
  factoryResetSpace as factoryResetSpaceSdk,
  getEncryptionState as getEncryptionStateSdk,
  lockEncryptionSession as lockEncryptionSessionSdk,
  unlockEncryptionSession as unlockEncryptionSessionSdk,
  unlockSpaceWithPassphrase as unlockSpaceWithPassphraseSdk,
  verifyKeychainAccess as verifyKeychainAccessSdk,
} from '@/api/generated/sdk.gen'
import { daemonClient } from './client'

// ── Response types ─────────────────────────────────────────────

/**
 * Encryption state as returned by `GET /encryption/state`.
 *
 * 加密状态，由 `GET /encryption/state` 返回。
 *
 * Field names are camelCase to match daemon serde `rename_all = "camelCase"`.
 */
export interface EncryptionStateResponse {
  /** Whether encryption has been initialized (passphrase configured). */
  initialized: boolean
  /** Whether the encryption session is currently unlocked and ready. */
  sessionReady: boolean
}

// ── Public API ─────────────────────────────────────────────────

/**
 * Fetch the current encryption state from the daemon.
 *
 * 从 daemon 获取当前加密状态。
 *
 * @returns Encryption initialization and session readiness.
 * @throws {DaemonApiError} On HTTP or session errors.
 */
export async function getEncryptionState(): Promise<EncryptionStateResponse> {
  // Route through the generated SDK; `callSdk` unwraps the SDK's outer `{ data }`
  // to the `EncryptionStateEnvelope`, and we unwrap `.data` to the payload. The
  // generated `EncryptionStateResponse` DTO is structurally equivalent to the
  // hand-written interface, bridged here to keep the public return type stable.
  const envelope = await daemonClient.callSdk(() => getEncryptionStateSdk({ throwOnError: true }))
  return envelope.data as unknown as EncryptionStateResponse
}

/**
 * Auto-unlock the encryption session via keyring.
 *
 * 通过 keyring 自动解锁加密会话（无需 passphrase）。
 *
 * Uses daemon HTTP API: `POST /encryption/unlock`
 *
 * @throws {DaemonApiError} On unlock errors (500) or if encryption not initialized.
 */
export async function unlockEncryption(): Promise<void> {
  // Action endpoint: the daemon auto-unlocks via keyring (no passphrase body).
  // The `{ success }` payload is discarded; we only need to surface failures.
  await daemonClient.callSdk(() => unlockEncryptionSessionSdk({ throwOnError: true }))
}

/**
 * Lock the encryption session, clearing the master key from memory.
 *
 * 锁定加密会话，从内存中清除主密钥。
 *
 * @throws {DaemonApiError} On HTTP or session errors.
 */
export async function lockEncryption(): Promise<void> {
  // Action endpoint: the `{ success }` payload is discarded; only failures matter.
  await daemonClient.callSdk(() => lockEncryptionSessionSdk({ throwOnError: true }))
}

/**
 * Verify macOS Keychain "Always Allow" permission for this app.
 *
 * 验证此应用的 macOS Keychain "始终允许" 权限。
 *
 * Uses daemon HTTP API: `GET /encryption/keychain-access`
 *
 * @returns True if permission is granted.
 * @throws {DaemonApiError} On HTTP or permission check errors.
 */
export async function verifyKeychainAccess(): Promise<boolean> {
  // Route through the generated SDK; `callSdk` unwraps to the
  // `KeychainAccessEnvelope`, then `.data.granted` is the boolean payload.
  const envelope = await daemonClient.callSdk(() => verifyKeychainAccessSdk({ throwOnError: true }))
  return envelope.data.granted
}

/**
 * Silent keyring-only resume (no passphrase) via `POST /encryption/unlock`.
 *
 * 用 keyring 已存 KEK 静默解锁。
 *
 * @returns `true` if the session resumed from the keyring, `false` if there is
 *          nothing to resume (no setup yet / keyslot missing).
 * @throws {DaemonApiError} on keyring↔keyslot drift / unexpected failure (500),
 *          so the caller can fall back to prompting for the passphrase.
 */
export async function trySilentUnlock(): Promise<boolean> {
  const envelope = await daemonClient.callSdk(() =>
    unlockEncryptionSessionSdk({ throwOnError: true })
  )
  return envelope.data.success
}

/**
 * User-driven passphrase unlock via `POST /encryption/unlock-with-passphrase`
 * (ADR-008 D15 — passphrase now rides the session-gated loopback API).
 *
 * 用户主动输入明文口令解锁。
 *
 * @returns The unlocked space id.
 * @throws {DaemonApiError} whose `.details.code` carries the semantic unlock
 *          error tag (`WRONG_PASSPHRASE`, `SPACE_NOT_INITIALIZED`, …) for the
 *          `security.ts` classifier to translate.
 */
export async function unlockSpaceWithPassphrase(passphrase: string): Promise<{ spaceId: string }> {
  const envelope = await daemonClient.callSdk(() =>
    unlockSpaceWithPassphraseSdk({ body: { passphrase }, throwOnError: true })
  )
  return { spaceId: envelope.data.spaceId }
}

/**
 * Factory-reset the space via `POST /encryption/factory-reset` — wipe key
 * material + clear setup status + cancel pending invitations.
 *
 * 重置并重新开始。
 *
 * @throws {DaemonApiError} whose `.details.code` carries the semantic reset
 *          error tag (`KEY_MATERIAL_WIPE_FAILED`, `STORAGE_FAILED`, …).
 */
export async function factoryResetSpace(): Promise<void> {
  await daemonClient.callSdk(() => factoryResetSpaceSdk({ throwOnError: true }))
}

File: crates/uc-webserver/src/api/encryption.rs
//! HTTP route handlers for encryption state and session management endpoints.

use axum::extract::State;
use axum::http::StatusCode;
use axum::routing::{get, post};
use axum::{Json, Router};
use tokio::sync::broadcast::error::SendError;
use tracing::{debug, info, warn};
use uc_application::facade::space_setup::UnlockSpaceError;
use uc_application::facade::{FactoryResetError, UnlockSpaceInput};
use uc_daemon_contract::api::dto::envelope::ApiEnvelope;
use uc_daemon_contract::constants::{ws_event, ws_topic};
use utoipa;

use crate::api::dto::encryption::{
    EncryptionActionResponse, EncryptionSessionReadyPayload, EncryptionStateResponse,
    KeychainAccessResponse, UnlockSpaceRequest, UnlockSpaceResponse,
};
use crate::api::dto::error::{log_facade_failure, ApiError};
use crate::api::server::DaemonApiState;
use crate::api::types::DaemonWsEvent;

/// 把 encryption facade 的 anyhow 错误映射为 500 + 根因日志。
///
/// 与 `RosterError` / `SearchFacadeError` 不同，encryption facade 当前向 webserver
/// 暴露 `anyhow::Error` 而非 enum，所以 `error_variant` 退化为 `"call_failed"`，
/// 分桶由 `op` 维度承担（state / unlock / lock / verify_keychain_access）。
fn map_encryption_internal(op: &'static str, message: String) -> ApiError {
    let api = ApiError::internal(message);
    log_facade_failure("encryption", op, "call_failed", api.status, &api.message);
    api
}

pub fn router() -> Router<DaemonApiState> {
    Router::new()
        .route("/encryption/state", get(get_encryption_state_handler))
        .route("/encryption/unlock", post(unlock_handler))
        .route(
            "/encryption/unlock-with-passphrase",
            post(unlock_with_passphrase_handler),
        )
        .route("/encryption/lock", post(lock_handler))
        .route("/encryption/factory-reset", post(factory_reset_handler))
        .route(
            "/encryption/keychain-access",
            get(verify_keychain_access_handler),
        )
}

/// Map the typed [`FactoryResetError`] onto an [`ApiError`] whose `code` is the
/// SCREAMING_SNAKE tag the frontend `FactoryResetCommandError` union switches on
/// (read off `DaemonApiError.details.code`). All variants are infra failures →
/// 500 with the cause string preserved in `message` (never a secret).
fn map_factory_reset_err(err: FactoryResetError) -> ApiError {
    use FactoryResetError as E;
    let (variant, code, message) = match err {
        E::KeyMaterialWipeFailed(msg) => {
            ("key_material_wipe_failed", "KEY_MATERIAL_WIPE_FAILED", msg)
        }
        E::StorageFailed(msg) => ("storage_failed", "STORAGE_FAILED", msg),
        E::Internal(msg) => ("internal", "INTERNAL", msg),
    };
    let api = ApiError {
        status: StatusCode::INTERNAL_SERVER_ERROR,
        code: code.to_string(),
        message,
        details: None,
    };
    log_facade_failure(
        "space_setup",
        "factory_reset",
        variant,
        api.status,
        &api.message,
    );
    api
}

/// Map the typed [`UnlockSpaceError`] onto an [`ApiError`] whose `code` is the
/// SCREAMING_SNAKE semantic tag the frontend `UnlockSpaceCommandError` union
/// switches on (read off `DaemonApiError.details.code` after `callSdk`
/// normalization). Statuses avoid `401` so `callSdk` does not trigger a
/// spurious session refresh + retry on a user-recoverable error.
fn map_unlock_err(err: UnlockSpaceError) -> ApiError {
    use UnlockSpaceError as E;
    let (variant, api): (&'static str, ApiError) = match err {
        E::SetupNotCompleted => (
            "setup_not_completed",
            ApiError {
                status: StatusCode::CONFLICT,
                code: "SETUP_NOT_COMPLETED".to_string(),
                message: "setup has not been completed".to_string(),
                details: None,
            },
        ),
        E::SpaceNotInitialized => (
            "space_not_initialized",
            ApiError {
                status: StatusCode::CONFLICT,
                code: "SPACE_NOT_INITIALIZED".to_string(),
                message: "space is not initialized on this device".to_string(),
                details: None,
            },
        ),
        E::WrongPassphrase => (
            "wrong_passphrase",
            ApiError {
                status: StatusCode::FORBIDDEN,
                code: "WRONG_PASSPHRASE".to_string(),
                message: "wrong passphrase".to_string(),
                details: None,
            },
        ),
        E::CorruptedKeyMaterial => (
            "corrupted_key_material",
            ApiError {
                status: StatusCode::UNPROCESSABLE_ENTITY,
                code: "CORRUPTED_KEY_MATERIAL".to_string(),
                message: "space key material is corrupted".to_string(),
                details: None,
            },
        ),
        // `msg` is an infra/migration string (never the passphrase) — safe to
        // surface to the 5xx root-cause log below.
        E::Internal(msg) => (
            "internal",
            ApiError {
                status: StatusCode::INTERNAL_SERVER_ERROR,
                code: "INTERNAL".to_string(),
                message: msg,
                details: None,
            },
        ),
    };
    log_facade_failure(
        "space_setup",
        "unlock_space",
        variant,
        api.status,
        &api.message,
    );
    api
}

/// GET /encryption/state
/// Returns the current encryption state and session readiness.
#[utoipa::path(
    get,
    path = "/encryption/state",
    operation_id = "getEncryptionState",
    tag = "encryption",
    responses(
        (status = 200, description = "Encryption state retrieved", body = EncryptionStateEnvelope),
        (status = 500, description = "Internal server error", body = ApiErrorResponse),
    )
)]
async fn get_encryption_state_handler(
    State(state): State<DaemonApiState>,
) -> Result<Json<ApiEnvelope<EncryptionStateResponse>>, ApiError> {
    let app = state.app_facade_or_error()?;
    let view = app
        .encryption
        .state()
        .await
        .map_err(|e| map_encryption_internal("encryption_state", e.to_string()))?;

    Ok(Json(ApiEnvelope::now(EncryptionStateResponse {
        initialized: view.initialized,
        session_ready: view.session_ready,
    })))
}

/// POST /encryption/unlock
/// Attempts to auto-unlock the encryption session using keyring-stored KEK.
/// No passphrase is required — credentials are retrieved from the OS keychain.
/// On success, broadcasts the `encryption.session_ready` WebSocket event.
#[utoipa::path(
    post,
    path = "/encryption/unlock",
    operation_id = "unlockEncryptionSession",
    tag = "encryption",
    responses(
        (status = 200, description = "Encryption session unlocked (or already ready)", body = EncryptionActionEnvelope),
        (status = 500, description = "Internal server error", body = ApiErrorResponse),
    )
)]
async fn unlock_handler(
    State(state): State<DaemonApiState>,
) -> Result<Json<ApiEnvelope<EncryptionActionResponse>>, ApiError> {
    let app = state.app_facade_or_error()?;

    // Route to `space_setup.try_resume_session()` rather than the bare
    // `encryption.unlock()`. The thin variant only resumes the session;
    // the SpaceSetupFacade variant also runs the switch-space migration
    // recovery hook (`resume_pending`) so a pending HandshakeDone replay
    // gets advanced the moment the session unlocks. Without that, a
    // device that crashed mid-`switch_space` would silently land here,
    // get a new master_key, leave the main table inline_data encrypted
    // with the previous key, and surface as "no history" in the UI —
    // exactly the wedged state we just dug out of on fedora dev.
    match app.try_resume_session().await {
        Ok(true) => {
            info!("encryption session auto-unlocked via keyring");

            let ts = chrono::Utc::now().timestamp_millis();
            let event_payload = EncryptionSessionReadyPayload { ts };
            let event = DaemonWsEvent {
                topic: ws_topic::ENCRYPTION.to_string(),
                event_type: ws_event::ENCRYPTION_SESSION_READY.to_string(),
                session_id: None,
                ts,
                payload: serde_json::to_value(&event_payload).unwrap_or(serde_json::Value::Null),
            };
            if let Err(SendError(_)) = state.event_tx.send(event) {
                debug!(
                    "failed to broadcast encryption.session_ready event — no active subscribers"
                );
            }

            Ok(Json(ApiEnvelope::now(EncryptionActionResponse {
                success: true,
            })))
        }
        Ok(false) => {
            info!("encryption not initialized, skipping auto-unlock");
            Ok(Json(ApiEnvelope::now(EncryptionActionResponse {
                success: false,
            })))
        }
        Err(e) => Err(map_encryption_internal(
            "encryption_unlock",
            format!("auto-unlock failed: {e}"),
        )),
    }
}

/// POST /encryption/unlock-with-passphrase
/// Unlocks the space with a user-supplied plaintext passphrase (ADR-008 D15).
///
/// Routes through `SpaceSetupFacade::unlock_space`, which (unlike the bare
/// `encryption.unlock`) also runs the switch-space migration recovery hook —
/// the same reason the keyring `unlock_handler` above delegates to
/// `try_resume_session`. On success it broadcasts `encryption.session_ready`
/// so WS subscribers react identically regardless of which unlock path ran.
///
/// D14: this endpoint is session-JWT gated (it is NOT in `PUBLIC_PATHS`) and
/// the handler MUST NOT log the request body — there is intentionally no
/// `?req` / passphrase field on any span or tracing event here.
#[utoipa::path(
    post,
    path = "/encryption/unlock-with-passphrase",
    operation_id = "unlockSpaceWithPassphrase",
    tag = "encryption",
    request_body = UnlockSpaceRequest,
    responses(
        (status = 200, description = "Space unlocked", body = UnlockSpaceEnvelope),
        (status = 403, description = "Wrong passphrase", body = ApiErrorResponse),
        (status = 409, description = "Setup not completed / space not initialized", body = ApiErrorResponse),
        (status = 422, description = "Space key material corrupted", body = ApiErrorResponse),
        (status = 500, description = "Internal server error", body = ApiErrorResponse),
    )
)]
async fn unlock_with_passphrase_handler(
    State(state): State<DaemonApiState>,
    Json(req): Json<UnlockSpaceRequest>,
) -> Result<Json<ApiEnvelope<UnlockSpaceResponse>>, ApiError> {
    let app = state.app_facade_or_error()?;
    let facade = app
        .space_setup
        .get()
        .cloned()
        .ok_or_else(|| ApiError::service_unavailable("space setup facade not assembled"))?;

    let result = facade
        .unlock_space(UnlockSpaceInput {
            passphrase: req.passphrase,
        })
        .await
        .map_err(map_unlock_err)?;

    info!("space unlocked via passphrase");

    let ts = chrono::Utc::now().timestamp_millis();
    let event_payload = EncryptionSessionReadyPayload { ts };
    let event = DaemonWsEvent {
        topic: ws_topic::ENCRYPTION.to_string(),
        event_type: ws_event::ENCRYPTION_SESSION_READY.to_string(),
        session_id: None,
        ts,
        payload: serde_json::to_value(&event_payload).unwrap_or(serde_json::Value::Null),
    };
    if let Err(SendError(_)) = state.event_tx.send(event) {
        warn!("failed to broadcast encryption.session_ready event — no active subscribers");
    }

    Ok(Json(ApiEnvelope::now(UnlockSpaceResponse {
        space_id: result.space_id.to_string(),
    })))
}

/// POST /encryption/lock
/// Locks the encryption session by clearing the master key.
#[utoipa::path(
    post,
    path = "/encryption/lock",
    operation_id = "lockEncryptionSession",
    tag = "encryption",
    responses(
        (status = 200, description = "Encryption session locked", body = EncryptionActionEnvelope),
        (status = 500, description = "Internal server error", body = ApiErrorResponse),
    )
)]
async fn lock_handler(
    State(state): State<DaemonApiState>,
) -> Result<Json<ApiEnvelope<EncryptionActionResponse>>, ApiError> {
    let app = state.app_facade_or_error()?;
    app.encryption.lock().await.map_err(|e| {
        map_encryption_internal("encryption_lock", format!("failed to lock encryption: {e}"))
    })?;

    info!("encryption session cleared (locked)");
    Ok(Json(ApiEnvelope::now(EncryptionActionResponse {
        success: true,
    })))
}

/// POST /encryption/factory-reset
/// Wipes key material + clears setup status + cancels pending invitations
/// (ADR-008 P3-1 / D15). Routes through `SpaceSetupFacade::factory_reset`,
/// mirroring the former in-process `factory_reset_space` Tauri command.
#[utoipa::path(
    post,
    path = "/encryption/factory-reset",
    operation_id = "factoryResetSpace",
    tag = "encryption",
    responses(
        (status = 200, description = "Space reset to factory state", body = EncryptionActionEnvelope),
        (status = 500, description = "Internal server error", body = ApiErrorResponse),
    )
)]
async fn factory_reset_handler(
    State(state): State<DaemonApiState>,
) -> Result<Json<ApiEnvelope<EncryptionActionResponse>>, ApiError> {
    let app = state.app_facade_or_error()?;
    let facade = app
        .space_setup
        .get()
        .cloned()
        .ok_or_else(|| ApiError::service_unavailable("space setup facade not assembled"))?;

    facade
        .factory_reset()
        .await
        .map_err(map_factory_reset_err)?;

    info!("space factory-reset completed");
    Ok(Json(ApiEnvelope::now(EncryptionActionResponse {
        success: true,
    })))
}

/// GET /encryption/keychain-access
/// Verifies macOS Keychain "Always Allow" permission for this app.
/// Returns `granted: true` if Keychain access succeeds silently, `false` if permission denied.
#[utoipa::path(
    get,
    path = "/encryption/keychain-access",
    operation_id = "verifyKeychainAccess",
    tag = "encryption",
    responses(
        (status = 200, description = "Keychain access verified", body = KeychainAccessEnvelope),
        (status = 500, description = "Internal server error", body = ApiErrorResponse),
    )
)]
async fn verify_keychain_access_handler(
    State(state): State<DaemonApiState>,
) -> Result<Json<ApiEnvelope<KeychainAccessResponse>>, ApiError> {
    let app = state.app_facade_or_error()?;
    let granted = app.encryption.verify_keychain_access().await.map_err(|e| {
        map_encryption_internal(
            "verify_keychain_access",
            format!("keychain access check failed: {e}"),
        )
    })?;

    Ok(Json(ApiEnvelope::now(KeychainAccessResponse { granted })))
}

#[cfg(test)]
mod tests {
    use super::*;

    /// The frontend `UnlockSpaceCommandError` union switches on the SCREAMING_SNAKE
    /// `code` (read off `DaemonApiError.details.code` after `callSdk` normalization),
    /// and `callSdk` would fire a spurious session refresh + retry on a `401`. So
    /// every user-recoverable variant must carry its semantic code and a non-401
    /// status.
    #[test]
    fn map_unlock_err_assigns_semantic_codes_and_avoids_401() {
        let cases = [
            (
                UnlockSpaceError::SetupNotCompleted,
                StatusCode::CONFLICT,
                "SETUP_NOT_COMPLETED",
            ),
            (
                UnlockSpaceError::SpaceNotInitialized,
                StatusCode::CONFLICT,
                "SPACE_NOT_INITIALIZED",
            ),
            (
                UnlockSpaceError::WrongPassphrase,
                StatusCode::FORBIDDEN,
                "WRONG_PASSPHRASE",
            ),
            (
                UnlockSpaceError::CorruptedKeyMaterial,
                StatusCode::UNPROCESSABLE_ENTITY,
                "CORRUPTED_KEY_MATERIAL",
            ),
        ];
        for (err, status, code) in cases {
            let api = map_unlock_err(err);
            assert_eq!(api.status, status);
            assert_eq!(api.code, code);
            assert_ne!(api.status, StatusCode::UNAUTHORIZED);
        }
    }

    #[test]
    fn map_unlock_err_internal_is_500_and_keeps_message() {
        let api = map_unlock_err(UnlockSpaceError::Internal(
            "migration resume failed: boom".to_string(),
        ));
        assert_eq!(api.status, StatusCode::INTERNAL_SERVER_ERROR);
        assert_eq!(api.code, "INTERNAL");
        assert_eq!(api.message, "migration resume failed: boom");
    }

    /// Factory-reset variants keep the `FactoryResetCommandError` semantic codes
    /// (frontend reads `DaemonApiError.details.code`) and preserve the infra
    /// cause string in `message`.
    #[test]
    fn map_factory_reset_err_keeps_semantic_codes_and_message() {
        let cases = [
            (
                FactoryResetError::KeyMaterialWipeFailed("disk i/o".to_string()),
                "KEY_MATERIAL_WIPE_FAILED",
                "disk i/o",
            ),
            (
                FactoryResetError::StorageFailed("settings db locked".to_string()),
                "STORAGE_FAILED",
                "settings db locked",
            ),
            (
                FactoryResetError::Internal("oops".to_string()),
                "INTERNAL",
                "oops",
            ),
        ];
        for (err, code, message) in cases {
            let api = map_factory_reset_err(err);
            assert_eq!(api.status, StatusCode::INTERNAL_SERVER_ERROR);
            assert_eq!(api.code, code);
            assert_eq!(api.message, message);
        }
    }
}
